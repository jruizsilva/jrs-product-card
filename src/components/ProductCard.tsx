import React, { createContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';
import {
  ProductContextProps,
  Product,
  onChangeArgs,
  InitialValue,
  ChildrenInterface,
} from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  // children?: ReactElement | ReactElement[]
  children: (args: ChildrenInterface) => JSX.Element;
  product: Product;
  className?: string;
  style?: CSSProperties;
  value?: number;
  initialValue: InitialValue;
  onChange?: (args: onChangeArgs) => void;
}

const ProductCard = ({
  product,
  children,
  className,
  style,
  value,
  onChange,
  initialValue,
}: Props) => {
  const {
    counter,
    increaseBy,
    reset,
    isMaxCountReached,
    maxCount,
  } = useProduct({
    onChange,
    product,
    value,
    initialValue,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
        isMaxCountReached,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          product,
          increaseBy,
          isMaxCountReached,
          maxCount,
          reset,
        })}
      </div>
    </Provider>
  );
};

export default ProductCard;
