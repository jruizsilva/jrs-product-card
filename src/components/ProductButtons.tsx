import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface Props {
  className?: string;
  style?: CSSProperties;
}

const ProductButtons = ({ className, style }: Props) => {
  const { counter, increaseBy, isMaxCountReached } = useContext(ProductContext);

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        disabled={isMaxCountReached}
        className={styles.buttonAdd}
        onClick={() => increaseBy(1)}
      >
        +
      </button>
    </div>
  );
};

export default ProductButtons;
