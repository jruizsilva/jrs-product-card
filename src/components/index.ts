import { ProductCardHOCProps } from '../interfaces/interfaces';
import ProductButtons from './ProductButtons';
import ProductCardHOC from './ProductCard';
import ProductImage from './ProductImage';
import ProductTitle from './ProductTitle';

const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});

export { ProductCard, ProductButtons, ProductImage, ProductTitle };
