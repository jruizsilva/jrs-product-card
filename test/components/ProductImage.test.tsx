import React from 'react';
import { create } from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';
import { product2 } from '../data/products';

describe('<ProductImage />', () => {
  test('Debe mostrar el componente con la imagen pasada en las props', () => {
    const wrapper = create(<ProductImage img={product2.img} />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Debe mostrar una imagen por defecto obtenida del componente <ProductCard/>', () => {
    const wrapper = create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
