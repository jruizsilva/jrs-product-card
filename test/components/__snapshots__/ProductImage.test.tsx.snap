// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductImage /> Debe mostrar el componente con la imagen pasada en las props 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="./coffee-mug.png"
/>
`;

exports[`<ProductImage /> Debe mostrar una imagen por defecto obtenida del componente <ProductCard/> 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
